File: Java/LinkedList/01_Singly_Linear_LinkedList_Java/SinglyLL_001.java
/*
===============================================================================

Program Name    : Singly Linear Linked List - Node Creation
Program Number  : 001
Description     : This program demonstrates the creation of a single node
                  for a Singly Linear Linked List in Java.
                  A node object is created dynamically using the
                  'new' keyword. The data member is initialized with
                  an integer value, while the next reference is set
                  to null, indicating that the node is currently
                  not connected to any other node.
Date            : July 2026
Language        : Java
Author          : Ishwari Vijaykumar Surve

===============================================================================

Concepts Used:

1. Classes and Objects
2. Self-Referential Class
3. Object References
4. Dynamic Object Creation using new
5. Instance Variables
6. Reference Initialization
7. Basic Memory Representation

Key Learning Points:

- Every node is an object in Java.
- Objects are created dynamically using the new keyword.
- The next reference stores the address of the next node object.
- A null reference indicates the end of the linked list.
- This program forms the basic building block of a Singly Linear Linked List.

===============================================================================
*/

class node
{
    public int data;
    public node next;
}

class SinglyLL001
{
    public static void main(String A[])
    {
        node newn = null;

        newn = new node();

        newn.data = 11;
        newn.next = null;

        System.out.println(newn.data);
        System.out.println(newn.next);
    }
}

/*
===============================================================================
Expected Output
===============================================================================

11
null

===============================================================================
Memory Diagram
===============================================================================

             newn
               |
               v
      +------------------+
      |    data = 11     |
      |------------------|
      |    next = null   |
      +------------------+

===============================================================================

Step-by-Step Execution

===============================================================================

Step 1:
A reference variable named newn is declared.

    node newn = null;

Currently it does not refer to any object.

Step 2:
Memory for a node object is allocated using the new keyword.

    newn = new node();

The reference variable now points to the newly created object.

Step 3:
The data member is initialized.

    newn.data = 11;

Step 4:
The next reference is initialized.

    newn.next = null;

Since there is no next node, the reference remains null.

Step 5:
The data member is displayed.

Output:

11

Step 6:
The next reference is displayed.

Output:

null

===============================================================================
Object Representation
===============================================================================

        newn
          |
          v

   +------------------+
   |   data : 11      |
   |------------------|
   |   next : null    |
   +------------------+

===============================================================================
Operations Performed
===============================================================================

1. Created a self-referential class.
2. Declared an object reference.
3. Allocated memory using the new keyword.
4. Stored an integer value in the node.
5. Initialized the next reference with null.
6. Displayed the node contents.

===============================================================================
Time Complexity
===============================================================================

Object Creation      : O(1)
Data Assignment      : O(1)
Display Values       : O(1)

Space Complexity     : O(1)


===============================================================================
Conclusion
===============================================================================

This program demonstrates the creation of a single node for a
Singly Linear Linked List using Java. It introduces the concepts
of classes, objects, object references, and dynamic object creation.
Understanding node creation is the first step toward implementing
complete linked list operations such as insertion, deletion,
traversal, and searching.

===============================================================================
*/

File: C++/LinkedList/01_Singly_Linear_LinkedList_CPP/SinglyLL_001.cpp
/*
===============================================================================
Program Name    : Singly Linked List - Node Structure and Memory Size
Program Number  : 001  
Description     : This program introduces the basic node structure used in a
                  Singly Linear Linked List. It demonstrates how a self-
                  referential structure is created and displays the memory
                  occupied by a single node using the sizeof operator.

Date            : July 2026
Language        : C++
Author          : Ishwari Vijaykumar Surve
===============================================================================

Concepts Used:
1. Structure
2. Self-Referential Structure
3. Pointer Declaration
4. typedef
5. sizeof Operator
6. Structure Packing using #pragma pack(1)
7. Node Representation

Key Learning Points:
- A node contains data and a pointer to the next node.
- next stores the address of the next node.
- #pragma pack(1) removes padding bytes.
- sizeof() returns the total memory occupied by a node.
- typedef improves code readability.

===============================================================================
*/

#include<iostream>
using namespace std;

#pragma pack(1)

struct node
{
    int data;
    struct node *next;
};

typedef struct node NODE;
typedef struct node *PNODE;
typedef struct node **PPNODE;

int main()
{
    NODE obj;

    cout<<"Size of node is : "<<sizeof(obj)<<endl;

    return 0;
}

/*
===============================================================================
Expected Output:
===============================================================================

Size of node is : 12

(Note: Output may vary depending on compiler, system architecture,
and structure packing.)

===============================================================================
Memory Layout:
===============================================================================

Node Structure

+------------+----------------------+
| data (4B)  | next pointer (8B)    |
+------------+----------------------+

Total Size (with #pragma pack(1)) = 12 Bytes

Without #pragma pack(1):

+------------+---------+----------------------+
| data (4B)  | Padding | next pointer (8B)    |
+------------+---------+----------------------+

Total Size = 16 Bytes (Compiler Dependent)

===============================================================================
Pointer Explanation:
===============================================================================

struct node
{
    int data;
    struct node *next;
};

data
└── Stores the integer value.

next
└── Stores the address of the next node.

typedef Usage:

NODE
└── Represents struct node

PNODE
└── Pointer to a node

PPNODE
└── Pointer to pointer to a node

===============================================================================
Program Flow:
===============================================================================

Create Node Object
        │
        ▼
Calculate sizeof(node)
        │
        ▼
Display Node Size
        │
        ▼
Program Ends

===============================================================================
Time Complexity:
===============================================================================

sizeof()          : O(1)
Space Complexity  : O(1)

===============================================================================
Conclusion:
===============================================================================

Program 001 introduces the basic node structure required for implementing a
Singly Linked List in C++. It demonstrates the use of a self-referential
structure, pointer declarations, typedef, and the sizeof operator to understand
the memory occupied by a node. This program serves as the foundation for all
subsequent linked list operations.

===============================================================================
*/
